### PR TITLE
BUG: Lazy cube with mask not saved correctly to NetCDF

### PR DESCRIPTION
**Pull request by cpelley**
Currently the `fill_value` keyword argument to [createVariable](https://github.com/SciTools/iris/blob/7890a7292d988e3ce2fbf31559fcb5ba39dbb041/lib/iris/fileformats/netcdf.py#L1447) is not utilised
when writing a NetCDF file when the data is lazy.

Determining the `fill_value` is not straight forward as biggus returns a default value (not None) so how do we differentiate between an ordinary array (pass a `fill_value=None`) and a masked array (pass a  `fill_value = cube.lazy_data().fill_value`) without touching the data??

Here is a simple illustration of the problem if my added integration test is not good enough:
http://nbviewer.ipython.org/gist/cpelley/2da0b8dd347f7a4772e0

### CODE REVIEW

**Comment by cpelley** [comment]
Further to the above, when not utilising a specific `fill_value` (which is a convention attribute) the behaviour defaults to the default NetCDF 'fill value' for the corresponding type.

**Comment by marqh** [comment]
Hello @cpelley 

Please can you clarify whether the changes in this PR represent all the changes needed to meet the use case, or whether there are a set of capabilities which you require, but are not met by this change?

thank you

**Comment by cpelley** [comment]
> Hello @cpelley
Please can you clarify whether the changes in this PR represent all the changes needed to meet the use case, or whether there are a set of capabilities which you require, but are not met by this change?
thank you

Thank you for poking this @marqh, yes this should be all the necessary changes.  I initially put up the PR with an added integration test which fails to demonstrate the problem since I could not think of a good solution to the problem.  With no sparked interest I have since put up a simple solution which people may or may not be happy with.  The NetCDF saver now always assigns the fill value for lazy data.

i.e. `fill_value=cube.lazy_data().fill_value`

I see no downside to this especially since learning about the default NetCDF data type fill values, this simply makes it explicitly an attribute.

**Comment by pelson** [comment]
Ping @rhattersley - note: this has been put against v1.8.x.

**Comment by marqh** [comment]
I understand the problem being addressed here and I can see that the solution presented delivers.

I would like to be confident that this is not changing the behaviour in the case where a cube does not have masked data.

```
>>> cube = iris.cube.Cube(np.arange(4))
>>> cube.add_dim_coord(iris.coords.DimCoord(np.arange(4), long_name='bla'), 0)
>>> iris.save(cube, 'atmp.nc', unlimited_dimensions=[])
>>> ncube = iris.load_cube('atmp.nc')
>>> print ncube.lazy_data().fill_value
-9223372036854775806

```

however, looking at the atmp.nc file:

```
netcdf atmp {
dimensions:
	bla = 4 ;
variables:
	int64 unknown(bla) ;
	int64 bla(bla) ;
		bla:units = "1" ;
		bla:long_name = "bla" ;

// global attributes:
		:Conventions = "CF-1.5" ;
}
```

there is no `_FillValue` attribute set.

On this basis, I think this is an appropriate change, dealing with the `_FillValue` attribute where it exists and representing data as numpy arrays where there is no `_FillValue`

File: lib/iris/tests/results/netcdf/netcdf_save_hybrid_height.cdl [hunk 1]
```diff
@@ -6,6 +6,7 @@ dimensions:
 	model_level_number = 10 ;
 variables:
 	float air_potential_temperature(time, model_level_number, grid_latitude, grid_longitude) ;
+		air_potential_temperature:_FillValue = -1.073742e+09f ;
```
Comment (marqh) [inline lib/iris/tests/results/netcdf/netcdf_save_hybrid_height.cdl]: why has this new line appeared?

Comment (cpelley) [inline lib/iris/tests/results/netcdf/netcdf_save_hybrid_height.cdl]: This relates to your above comment where a biggus array always has a fill value.

**Comment by cpelley** [comment]
> On this basis, I think this is an appropriate change, dealing with the _FillValue attribute where it exists and representing data as numpy arrays where there is no _FillValue

Thank your for your interest in this @marqh.  I should probably mention that your test case utilises non-lazy data.  If your cube were lazy, then the fill value would be recorded in the NetCDF as a result of this PR irrespective of whether its masked or not (your case would not be affected as it is not lazy).

I would argue that there is no differentiation between masked arrays and non masked arrays in NetCDF.  Iris creates the distinction when it asks 'are there are any masked elements?' and returns you an unmasked array if there are no masked elements and a masked array if there are any masked elements.

This alters the existing behaviour of assigning an explicit fill value to the file, however, NetCDF always has a type specific fill value though before not an explicit attribute.  This PR would simply make it explicit for the deferred cases, but not alter the default interpretation of any files.

The alternative is to touch the data before saving to differentiate between masked and unmasked or to make significant changes to biggus with complex logic to cater for this.  I don't think the distinction between masked and unmasked is a NetCDF one, so I was happy with this simple solution.

**Comment by marqh** [comment]
If I understand this correctly, this is changing the behaviour of Iris such that is explicitly sets a fill_value every time it writes out data from a lazy_array.  If there is a fill_value to use, it uses it, otherwise it uses Biggus' default.
Currently this is not the case.
However, you are asserting that there is a default fill value implicit in netCDF so this is not semantic change to the data and metadata being provided, just a change to the explicit attribute metadata
 - is this a fair interpretation?

I would like to be confident that we are not altering the behaviour of such data when it is loaded.

If I load the result from one of the tests as saved before this PR and the same result file, but after this PR, are the Cubes the same?  Have I affected how they are interpreted by Iris?
I think I have.

**Comment by cpelley** [comment]
> However, you are asserting that there is a default fill value implicit in netCDF so this is not semantic change to the data and metadata being provided, just a change to the explicit attribute metadata
>    is this a fair interpretation?

Yes exactly and I have confirmed that Iris does comply with the default NetCDF fill_values.

> If I load the result from one of the tests as saved before this PR and the same result file, but after this PR, are the Cubes the same? Have I affected how they are interpreted by Iris?
I think I have.

For the case where we had lazy data, we threw away the fill value, this would mean that the 'masked' data values being loaded are now different (previously, the NetCDF dtype default fill_value value, now the fill_value value actually associated before saving).

Yes it can be considered a behavioural change but it is a bug.  This does not match the behaviour for non-lazy data.
I'm now leaning towards targeting iris master.  If we target iris master, I'll put a deprecation warning of the current behaviour targeted to 1.8.x.

**Comment by marqh** [comment]
I agree that losing the assigned _fill_value during a load and save which maintains the lazy data is an undesired behaviour, and i agree that this PR addresses that with minimal code changes; which is good.

i don't think this PR matches the behaviour for non-lazy and lazy data completely.  I'm not saying it has to, but it is worth a small consideration.

Given this PR: 
 - If I load a cube from a netCDF file variable with no _fill_value set, then save it, a _fill_value will be set
 - If I load a cube from a netCDF file variable with no _fill_value set, touch the data, then save it, no _fill_value will be set
 - If I load a cube from a netCDF file variable with a _fill_value set, then save it, that _fill_value will be preserved
 - If I load a cube from a netCDF file variable with a _fill_value set, touch the data, then save it, that _fill_value will be preserved


We can accept this behaviour, but I'd like to explicitly accept it

**Comment by rhattersley** [comment]
> If I load a cube from a netCDF file variable with a _fill_value set, touch the data, then save it, that _fill_value will be preserved

Only if the data include values which are equal to the fill value. Otherwise the _FillValue attribute will disappear.

**Comment by marqh** [comment]
> Only if the data include values which are equal to the fill value. Otherwise the _FillValue attribute will disappear.

interesting, I hadn't picked up on this.

So, if my data array is not a masked array, or if it is a masked array but there are no masked values, then no _FillValue attribute will be added, even if it is set by
```
self._dataset.createVariable(
...
fill_value=cube.lazy_data().fill_value,
...
```
during creation of the netcdf variable object

that seems to me to reduce the impact of this PR, making it easier to accept onto 1.8.x as a bug fix

**Comment by rhattersley** [comment]
### Yet Another Summary :wink:
*This time, focussed on the changes in behaviour introduced by this PR.*

For a file with an explicit `_FillValue` attribute:
* Irrespective of whether the data contain the missing value, it no longer loses its _FillValue when saved lazily. +1

For a file without an explicit `_FillValue` attribute:
* Irrespective of whether the data contain the *implicit* missing value, it gains an explicit _FillValue when saved lazily.
 * If the data contain the missing value this matches the behaviour when saved normally.  +0
 * If the data do not contain the missing value this does **not** match the behaviour when saved normally. Formally this is not a semantic change, but in the Real World :tm: ... ? -0

NB. We still lose the explicit _FillValue when the file does not contain any missing data (irrespective of lazy/normal and master vs. this PR). :-1: But arguably that's a problem with the data representation from *loading* a Cube so could be tackled separately.

On balance it looks like this PR is a step in the right direction and can arguably be classified as a bug fix.

**Comment by marqh** [comment]
This looks good to merge to me now, I think it is a valid bug fix for 1.8.0 and the implications are well understood
:+1:

**Comment by rhattersley** [comment]
Thanks @cpelley for fixing this, and @marqh for your careful review. As the polo-neck jazz guy from The Fast Show would say, "Nice!" :wink:

**Comment by cpelley** [comment]
Thanks both, appreciate your time and patience on this.

**Pull request merged**